kml/minzoom_quadtree.hpp

```diff
@@ -0,0 +1,177 @@
+    // level call of `traverse()` makes it one-based so it effectively became `2 - baseZoom`,
+    // which have exactly the same meaning as another "zoomLevel"s in the project.
+    // Scaling (enlarging) the size of bbox by `1 / scale` is required so that one can
+    // set arbitrary values for `countPerTile`, not only a powers of four.
+    auto const scale = std::frexp(spacing, &baseZoom);
+    baseZoom = 1 - baseZoom;
+
+    if (baseZoom >= maxZoom)
+      return;
+
+    auto const size = std::max(bbox.SizeX(), bbox.SizeY()) / scale;
+    bbox.SetSizes(size, size);
+
+    // Calculate coordinate along z-order curve.
+    auto const [minX, minY] = bbox.LeftBottom();
+    auto const gridScale = std::numeric_limits<uint32_t>::max() / size;
+    for (auto & elem : m_quadtree)
+    {
+      auto const [x, y] = elem.m_point;
+      auto const gridX = static_cast<uint32_t>(std::trunc((x - minX) * gridScale));
+      auto const gridY = static_cast<uint32_t>(std::trunc((y - minY) * gridScale));
+      // Sacrifice one of 32 possible levels for simplicity of traversing.
+      elem.m_zCurveCoord = bits::BitwiseMerge(gridX, gridY) >> 2;
+    }
+
+    std::sort(m_quadtree.begin(), m_quadtree.end());
+
+    auto const traverse = [&](auto const & traverse, auto const treeBeg, auto const treeEnd,
+                              uint64_t treeLevelMask, int zoom) -> Element *
+    {
+      if (treeBeg == treeEnd)
+        return nullptr;
+      if (std::next(treeBeg) == treeEnd)
+        return &*treeBeg;
+      if (++zoom >= maxZoom)
+      {
+        auto const rankLess = [&](auto const & lhs, auto const & rhs) -> bool
+        {
+          return m_less(lhs.m_value, rhs.m_value);
+        };
+        auto const result = std::max_element(treeBeg, treeEnd, rankLess);
+        auto const setMaxZoom = [&](auto & elem) { setMinZoom(elem.m_value, maxZoom); };
+        std::for_each(treeBeg, result, setMaxZoom);
+        std::for_each(std::next(result), treeEnd, setMaxZoom);
+        return &*result;
+      }
+
+      Element * topRanked = nullptr;
+      treeLevelMask >>= 2;
+      CHECK_NOT_EQUAL(treeLevelMask, 0, ());
+      uint64_t const quadIncrement = treeLevelMask & (treeLevelMask >> 1);
+      uint64_t quadIndex = 0;
+      auto const quadMaskEqual = [&](auto const & elem) -> bool
+      {
+        return (elem.m_zCurveCoord & treeLevelMask) == quadIndex;
+      };
+      auto quadBeg = treeBeg;
+      for (int quadrant = 0; quadrant < 4; ++quadrant)
+      {
+        ASSERT(std::is_partitioned(quadBeg, treeEnd, quadMaskEqual), ());
+        auto const quadEnd = std::partition_point(quadBeg, treeEnd, quadMaskEqual);
+        if (auto const elem = traverse(traverse, quadBeg, quadEnd, treeLevelMask, zoom))
+        {
+          if (topRanked == nullptr)
+            topRanked = elem;
+          else if (m_less(topRanked->m_value, elem->m_value))
+            setMinZoom(std::exchange(topRanked, elem)->m_value, std::max(zoom, 1));
```

Comment:
shouldn't it be ```zoom-1``` here?
what's the point of incrementing zoom as the very first mutating statement of this function? I think passing zoom+1 to recursive calls would be more readable

Reply:
> shouldn't it be zoom-1 here?

Not, because initially `baseZoom` is zero-based, whereas expected `m_minZoom` is one-based.

 > what's the point of incrementing zoom as the very first mutating statement of this function?

To make `zoom` one-based.

 > I think passing zoom+1 to recursive calls (thus having zoom constant within a single call's scope) would be more readable

I'll try to rewrite.